provider/pkg/gen/discovery.go

```diff
@@ -120,14 +139,94 @@ func findResourcesImpl(docName, parentName string, rest map[string]discovery.Res
 	return nil
 }
 
-func addFoundResource(resourceMap map[string]discoveryDocumentResource,typeName string, dd discoveryDocumentResource) {
+func addFoundResource(resourceMap map[string]discoveryDocumentResource, typeName string, dd discoveryDocumentResource) error {
 	for _, param := range dd.createMethod.Parameters {
 		if param.Location == "path" && isDeprecated(param.Description) {
 			// If a path parameter is deprecated, the URL is effectively deprecated, so skip this resource.
-			return
+			return nil
 		}
 	}
 
-	// TODO: resolve conflicts so that we never override resources with the same name
-	resourceMap[typeName] = dd
+	existing, has := resourceMap[typeName]
+	if !has {
+		// No conflict - use this resource.
+		resourceMap[typeName] = dd
+		return nil
+	}
+
+	// Check if two conflicting resources represent the same data type. Otherwise, error - we don't support this scenario.
+	if existing.createMethod.Response.Ref != dd.createMethod.Response.Ref {
+		return errors.Errorf("%q conflict: %q (%q) vs %q (%q)", typeName, existing.createMethod.Response.Ref, existing.createMethod.FlatPath, dd.createMethod.Response.Ref, dd.createMethod.FlatPath)
+	}
+
+	// Parse the paths of both resources to extract the parameters.
+	existingParams := findApiParams(existing)
+	newParams := findApiParams(dd)
+
+	// Check if one set of parameters is preferred over the other, pick the preferred one if so.
+	if preferParams(existingParams, newParams) {
+		return nil
+	}
+	if preferParams(newParams, existingParams) {
+		resourceMap[typeName] = dd
+		return nil
+	}
+
+	return errors.Errorf("%q incompatible params: %q vs %q", typeName, existing.createMethod.FlatPath, dd.createMethod.FlatPath)
+}
+
+func preferParams(set codegen.StringSet, other codegen.StringSet) bool {
+	if containsSet(set, other) {
+		return true
+	}
+
+	diff := diffSet(set, other)
+	if len(diff) == 1 && diff.Has("regionId") && set.Has("location") {
+		return true
+	}
+
+	return false
+}
+
+func containsSet(set codegen.StringSet, subset codegen.StringSet) bool {
```

Comment:
Would it make sense to add the contains and diff functions as methods to the StringSet?

Reply:
https://github.com/pulumi/pulumi/pull/7256